File: app/src/main/java/com/inappropirates/lightwalker/util/PropertyFormatter.java
package com.inappropirates.lightwalker.util;

import android.graphics.Color;

public class PropertyFormatter
{
    private final static int colorRoundValue = 17;

    public static String getStringVal(String key, Object value)
    {
        String stringVal = null;

        if (key.toLowerCase().contains("color"))
        {
            // in order to round things out a big so you can easily pick a solid green or yellow or whatever,
            // pick the color down to it's argb values, round those, and then put it back into a color
            Integer color = (Integer) value;
            int newColor = Color.argb(Color.alpha(color), RoundColor(Color.red(color)), RoundColor(Color.green(color)), RoundColor(Color.blue(color)));
            float[] hsvColor = new float[3];
            Color.colorToHSV(newColor, hsvColor);

            StringBuilder builder = new StringBuilder();
            builder.append((int) Util.map(hsvColor[0], 0, 360, 0, 255))
                    .append(",")
                    .append((int) Util.map(hsvColor[1], 0, 1, 0, 255))
                    .append(",")
                    .append((int) Util.map(hsvColor[2], 0, 1, 0, 255));
            stringVal = builder.toString();
        } else
        {
            if (value instanceof Integer)
                stringVal = value.toString();
            else if (value instanceof String)
                stringVal = (String) value;
            else if (value instanceof Boolean)
                stringVal = ((Boolean) value) ? "1" : "0";
        }

        return stringVal;
    }

    private static int RoundColor(int color)
    {
        int modValue = color % colorRoundValue;
        if (modValue == 0)
            return color;

        int cutoff = (int) (colorRoundValue * 0.8);
        if (modValue <= cutoff)
            return color - modValue;
        else
            return color + (colorRoundValue - modValue);

    }
}

File: app/src/main/kotlin/com/inappropirates/lightwalker/remote/util/PropertyFormatter.kt
package com.inappropirates.lightwalker.remote.util

import androidx.compose.ui.graphics.Color
import com.inappropirates.lightwalker.remote.ui.HSVColor
import com.inappropirates.lightwalker.remote.ui.fromHex

object PropertyFormatter {
    private const val COLOR_ROUND_VALUE = 17

    fun getStringVal(key: String, value: Any): String {
        val it = if (value is String && value.startsWith("#")) {
            HSVColor.fromAndroidColor(Color.fromHex(value))
        } else {
            value
        }

        return when (it) {
            is Int -> it.toString()
            is String -> it
            is Float -> "%.0f".format(it)
            is HSVColor -> {
                val builder = StringBuilder()
                builder.append(it.hue)
                    .append(",")
                    .append(it.sat)
                    .append(",")
                    .append(it.value)
                builder.toString()
            }

            is Boolean -> {
                if (it) "1" else "0"
            }

            else -> throw RuntimeException("Unhandled property type! key: $key value: $value type: ${value.javaClass}")
        }
    }

    private fun roundColor(color: Int): Int {
        val modValue = color % COLOR_ROUND_VALUE
        if (modValue == 0) return color

        val cutoff = (COLOR_ROUND_VALUE * 0.8).toInt()
        return if (modValue <= cutoff) color - modValue
        else color + (COLOR_ROUND_VALUE - modValue)
    }
}
